File: cpp/client/member.cc
#include <webcface/member.h>
#include <webcface/data.h>
#include <webcface/func.h>
#include <webcface/view.h>
#include <webcface/event_target.h>
#include <webcface/client_data.h>

namespace WebCFace {

Value Member::value(const std::string &field) const {
    return Value{*this, field};
}
Text Member::text(const std::string &field) const { return Text{*this, field}; }
Func Member::func(const std::string &field) const { return Func{*this, field}; }
View Member::view(const std::string &field) const { return View{*this, field}; }
Log Member::log() const { return Log{*this}; }

EventTarget<Value, std::string> Member::onValueEntry() const {
    return EventTarget<Value, std::string>{&dataLock()->value_entry_event,
                                           member_};
}
EventTarget<Text, std::string> Member::onTextEntry() const {
    return EventTarget<Text, std::string>{&dataLock()->text_entry_event,
                                          member_};
}
EventTarget<Func, std::string> Member::onFuncEntry() const {
    return EventTarget<Func, std::string>{&dataLock()->func_entry_event,
                                          member_};
}
EventTarget<View, std::string> Member::onViewEntry() const {
    return EventTarget<View, std::string>{&dataLock()->view_entry_event,
                                          member_};
}
EventTarget<Member, std::string> Member::onSync() const {
    return EventTarget<Member, std::string>{&dataLock()->sync_event, member_};
}

std::vector<Value> Member::values() const {
    auto keys = dataLock()->value_store.getEntry(*this);
    std::vector<Value> ret(keys.size());
    for (std::size_t i = 0; i < keys.size(); i++) {
        ret[i] = value(keys[i]);
    }
    return ret;
}
std::vector<Text> Member::texts() const {
    auto keys = dataLock()->text_store.getEntry(*this);
    std::vector<Text> ret(keys.size());
    for (std::size_t i = 0; i < keys.size(); i++) {
        ret[i] = text(keys[i]);
    }
    return ret;
}
std::vector<Func> Member::funcs() const {
    auto keys = dataLock()->func_store.getEntry(*this);
    std::vector<Func> ret(keys.size());
    for (std::size_t i = 0; i < keys.size(); i++) {
        ret[i] = func(keys[i]);
    }
    return ret;
}
std::vector<View> Member::views() const {
    auto keys = dataLock()->view_store.getEntry(*this);
    std::vector<View> ret(keys.size());
    for (std::size_t i = 0; i < keys.size(); i++) {
        ret[i] = view(keys[i]);
    }
    return ret;
}

} // namespace WebCFace

File: client_js/src/member.ts
import { Value, Text, Log } from "./data.js";
import { Func, FuncCallback, AnonymousFunc, Arg } from "./func.js";
import { View } from "./view.js";
import { Field } from "./field.js";
import { FieldWithEvent, eventType } from "./event.js";

export class Member extends Field {
  constructor(base: Field, member = "") {
    super(base.data, member || base.member_, "");
  }
  get name() {
    return this.member_;
  }
  value(name: string) {
    return new Value(this, name);
  }
  text(name: string) {
    return new Text(this, name);
  }
  view(name: string) {
    return new View(this, name);
  }
  func(name: string): Func;
  func(callback: FuncCallback, returnType: number, args: Arg[]): AnonymousFunc;
  func(...args: [string] | [FuncCallback, number, Arg[]]) {
    if (typeof args[0] === "string") {
      return new Func(this, args[0]);
    } else {
      return new AnonymousFunc(this, args[0], args[1] || 0, args[2] || []);
    }
  }
  log() {
    return new Log(this);
  }
  values() {
    return this.data.valueStore
      .getEntry(this.member_)
      .map((n) => this.value(n));
  }
  texts() {
    return this.data.textStore.getEntry(this.member_).map((n) => this.text(n));
  }
  views() {
    return this.data.viewStore.getEntry(this.member_).map((n) => this.view(n));
  }
  funcs() {
    return this.data.funcStore.getEntry(this.member_).map((n) => this.func(n));
  }
  get onValueEntry() {
    return new FieldWithEvent<Value>(
      eventType.valueEntry(this),
      this.data,
      this.member_
    );
  }
  get onTextEntry() {
    return new FieldWithEvent<Text>(
      eventType.textEntry(this),
      this.data,
      this.member_
    );
  }
  get onFuncEntry() {
    return new FieldWithEvent<Func>(
      eventType.funcEntry(this),
      this.data,
      this.member_
    );
  }
  get onViewEntry() {
    return new FieldWithEvent<View>(
      eventType.viewEntry(this),
      this.data,
      this.member_
    );
  }
  get onSync() {
    return new FieldWithEvent<Member>(
      eventType.sync(this),
      this.data,
      this.member_
    );
  }
}
